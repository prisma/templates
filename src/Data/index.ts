export * as Data from './index_'
