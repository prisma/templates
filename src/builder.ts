import { BaseTemplateParametersResolved, BuildTemplate, File, Template } from './types'
import { Index, mapValues } from './utils'

/**
 * Build the given template with given parameters.
 */
export const build = <T extends Template>(
  template: T,
  parameters: GlobalPrismaTemplateParameters[T['metadata']['name']]
): BuildTemplate<T> => {
  const parameters_ = {
    ...parameters,
    ...template.Parameters.defaults,
  }

  return {
    ...template,
    files: runFileTransformers(transformers, template.files, parameters_),
  }
}

// Template Transformers

const prismaSchemaProviderTransformer: FileTransformer = (params) => {
  const { file, parameters } = params

  let content = file.content

  if (file.path === 'prisma/schema.prisma') {
    content = content.replace(`provider = "postgresql"`, `provider = "${parameters.datasourceProvider}"`)

    if (parameters.datasourceProvider === 'sqlserver') {
      content = content.replace(
        `provider = "prisma-client-js"`,
        `provider = "prisma-client-js"
  previewFeatures = ["microsoftSqlServer"]`
      )
    }
  }

  return {
    ...file,
    content,
  }
}

const transformers = [prismaSchemaProviderTransformer]

// Template Transformers System

type TransformerParams = { file: File; parameters: BaseTemplateParametersResolved }

type FileTransformer = (params: TransformerParams) => File

const runFileTransformers = (
  transformers: FileTransformer[],
  files: Index<File>,
  parameters: TransformerParams['parameters']
) => {
  return mapValues(files, (file) => {
    return {
      ...file,
      content: transformers.reduce((file, transformer): File => {
        return transformer({
          file,
          parameters,
        })
      }, file),
    }
  })
}
