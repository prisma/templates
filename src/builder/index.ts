export * as Builder from './builder'
