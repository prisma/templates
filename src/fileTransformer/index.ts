export * as FileTransformer from './fileTransformer'
