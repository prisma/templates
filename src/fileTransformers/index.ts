export * as FileTransformers from './index_'
