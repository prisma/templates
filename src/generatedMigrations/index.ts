export * as MigrationsSql from './index_'