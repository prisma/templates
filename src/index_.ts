import { Generated } from './generated'

export * as $Types from './types'

export const Templates = Generated.Templates

export const Utils = {
  getVercelDeployButtonUrl(params: { repositoryOwner: string; repositoryHandle: string }) {
    const repository = encodeURIComponent(params.repositoryOwner + '/' + params.repositoryHandle)
    const envVars = ['DATABASE_URL', 'DATABASE_MIGRATE_URL'].join(',')
    const envVarDescription = encodeURIComponent(
      "Connection string for the database this deployment will talk to. If you're using the Prisma Data Proxy, then use its connection string."
    )

    return `https://vercel.com/new/import?repository-url=https%3A%2F%2Fgithub.com%2F${repository}&env=${envVars}&envDescription=${envVarDescription}`
  },
}
