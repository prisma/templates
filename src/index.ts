export * as PrismaTemplates from './index_'
import * as PrismaTemplates from './index_'

const x = new PrismaTemplates.Templates.Empty({})
