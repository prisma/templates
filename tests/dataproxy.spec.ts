import { PrismaTemplates } from '~/src'

Object.values(PrismaTemplates.Templates).forEach((Template) => {
  it.only(`${Template.metadata.displayName}`, () => {
    const template = new Template({
      datasourceProvider: 'mysql',
      repositoryOwner: 'prisma',
      repositoryHandle: 'templates-node',
      dataproxy: true,
    })
    expect(template.files['prisma/schema.prisma']).toMatchSnapshot()
  })
})

Object.values(PrismaTemplates.Templates).forEach((Template) => {
  it(`is enabled by default: ${Template.metadata.displayName}`, () => {
    expect(
      new Template({
        datasourceProvider: 'mysql',
        repositoryOwner: 'prisma',
        repositoryHandle: 'templates-node',
        dataproxy: true,
      })
    ).toEqual(
      new Template({
        datasourceProvider: 'mysql',
        repositoryOwner: 'prisma',
        repositoryHandle: 'templates-node',
      })
    )
  })
})

Object.values(PrismaTemplates.Templates).forEach((Template) => {
  it(`can be disabled: ${Template.metadata.displayName}`, () => {
    const template = new Template({
      datasourceProvider: 'mysql',
      repositoryOwner: 'prisma',
      repositoryHandle: 'templates-node',
      dataproxy: false,
    })
    expect(template.files['prisma/schema.prisma']).toMatchSnapshot()
  })
})
