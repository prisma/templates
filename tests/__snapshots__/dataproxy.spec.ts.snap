// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog 1`] = `
Object {
  "content": "generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model Post {
  post_id   Int     @id @default(autoincrement())
  content   String? @db.Text
  title     String @db.Text
  author_id Int?
  author    User?   @relation(fields: [author_id], references: [user_id])
}

model Profile {
  bio        String? @db.Text
  profile_id Int     @id @default(autoincrement())
  user_id    Int
  user       User    @relation(fields: [user_id], references: [user_id])
}

model User {
  email    String @db.Text    @unique
  name     String? @db.Text
  user_id  Int       @id @default(autoincrement())
  posts    Post[]
  profiles Profile[]
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`Empty 1`] = `
Object {
  "content": "// An empty Prisma Schema with no models, perfect as a starting point for your own schema.

datasource db {
provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

generator client {
  provider = \\"prisma-client-js\\"
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`Music Streaming Service 1`] = `
Object {
  "content": "// Schema for a fictional music streaming service
// This service contains songs, albums, artists & playlists, and tracks user interactions with each of them

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model User {
  id           String @db.Text        @id @default(uuid())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  name         String @db.Text
  email        String @db.Text        @unique
  interactions Interaction[]
  playlists    Playlist[]
}

model Interaction {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  songId    String? @db.Text
  userId    String? @db.Text
  isLiked   Boolean  @default(false)
  playCount Int
  song      Song?    @relation(fields: [songId], references: [id])
  user      User?    @relation(fields: [userId], references: [id])
}

model Song {
  id           String @db.Text        @id @default(uuid())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  name         String @db.Text
  albumId      String? @db.Text
  artistId     String @db.Text
  length       Float
  track        Int?
  lyrics       String? @db.Text
  fileUrl      String @db.Text
  album        Album?        @relation(fields: [albumId], references: [id])
  artist       Artist        @relation(fields: [artistId], references: [id])
  interactions Interaction[]
  playlist     Playlist[]
}

model Artist {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String @db.Text
  songs     Song[]
  albums    Album[]
}

model Album {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String @db.Text
  cover     String @db.Text
  songs     Song[]
  artists   Artist[]
}

model Playlist {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  userId    String? @db.Text
  name      String @db.Text
  user      User?    @relation(fields: [userId], references: [id])
  songs     Song[]
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`Rentals Platform 1`] = `
Object {
  "content": "// Schema for a fictional hotel
// This hotel tracks rooms, user reservations & reviews

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model User {
  id           String @db.Text        @id @default(uuid())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  email        String @db.Text        @unique
  name         String @db.Text
  reservations Reservation[]
  reviews      Review[]
}

model Reservation {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  userId    String? @db.Text
  startDate DateTime
  endDate   DateTime
  price     Int
  total     Int
  roomId    String? @db.Text
  room      Room?    @relation(fields: [roomId], references: [id])
  user      User?    @relation(fields: [userId], references: [id])
}

model Room {
  id                 String @db.Text        @id @default(uuid())
  createdAt          DateTime      @default(now())
  updatedAt          DateTime      @updatedAt
  totalOccupancy     Int           @default(5)
  totalBedrooms      Int           @default(5)
  totalBathrooms     Int           @default(5)
  summary            String? @db.Text
  address            String @db.Text
  isSmoking          Boolean       @default(false)
  hasTV              Boolean       @default(true)
  hasAirConditioning Boolean       @default(true)
  hasHeating         Boolean       @default(true)
  hasWIFI            Boolean       @default(true)
  price              Int
  media              Media[]
  reservations       Reservation[]
}

model Review {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  rating    Int
  comment   String @db.Text
  userId    String @db.Text
  user      User     @relation(fields: [userId], references: [id])
}

model Media {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  roomId    String? @db.Text
  fileName  String @db.Text
  room      Room?    @relation(fields: [roomId], references: [id])
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`SaaS 1`] = `
Object {
  "content": "// Schema for an identity service for a SaaS
// This service tracks existing users & their accounts, as well as invites for users who haven't joined yet

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model User {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String? @db.Text
  email     String @db.Text
  accountId String? @db.Text
  account   Account? @relation(fields: [accountId], references: [id])
}

model Account {
  id                   String @db.Text   @id @default(uuid())
  createdAt            DateTime @default(now())
  updatedAt            DateTime @updatedAt
  stripeCustomerId     String @db.Text
  stripeSubscriptionId String @db.Text
  referrer             String? @db.Text
  isActive             Boolean
  invites              Invite[]
  users                User[]
}

model Invite {
  id        String @db.Text   @id @default(uuid())
  dateSent  DateTime @default(now())
  email     String @db.Text
  accountId String? @db.Text
  isValid   Boolean  @default(true)
  account   Account? @relation(fields: [accountId], references: [id])
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`URL Shortener 1`] = `
Object {
  "content": "// Schema for a simple URL shortener
// This schema tracks short links by associating them to a user

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model Link {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  url       String @db.Text
  shortUrl  String @db.Text
  userId    String? @db.Text
  User      User?    @relation(fields: [userId], references: [id])
}

model User {
  id        String @db.Text   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String? @db.Text
  email     String @db.Text
  links     Link[]
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`can be disabled: Blog 1`] = `
Object {
  "content": "generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
  provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model Post {
  post_id   Int     @id @default(autoincrement())
  content   String?
  title     String
  author_id Int?
  author    User?   @relation(fields: [author_id], references: [user_id])
}

model Profile {
  bio        String?
  profile_id Int     @id @default(autoincrement())
  user_id    Int
  user       User    @relation(fields: [user_id], references: [user_id])
}

model User {
  email    String    @unique
  name     String?
  user_id  Int       @id @default(autoincrement())
  posts    Post[]
  profiles Profile[]
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`can be disabled: Empty 1`] = `
Object {
  "content": "// An empty Prisma Schema with no models, perfect as a starting point for your own schema.

datasource db {
  provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

generator client {
  provider = \\"prisma-client-js\\"
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`can be disabled: Music Streaming Service 1`] = `
Object {
  "content": "// Schema for a fictional music streaming service
// This service contains songs, albums, artists & playlists, and tracks user interactions with each of them

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
  provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model User {
  id           String        @id @default(uuid())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  name         String
  email        String        @unique
  interactions Interaction[]
  playlists    Playlist[]
}

model Interaction {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  songId    String?
  userId    String?
  isLiked   Boolean  @default(false)
  playCount Int
  song      Song?    @relation(fields: [songId], references: [id])
  user      User?    @relation(fields: [userId], references: [id])
}

model Song {
  id           String        @id @default(uuid())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  name         String
  albumId      String?
  artistId     String
  length       Float
  track        Int?
  lyrics       String?
  fileUrl      String
  album        Album?        @relation(fields: [albumId], references: [id])
  artist       Artist        @relation(fields: [artistId], references: [id])
  interactions Interaction[]
  playlist     Playlist[]
}

model Artist {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String
  songs     Song[]
  albums    Album[]
}

model Album {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String
  cover     String
  songs     Song[]
  artists   Artist[]
}

model Playlist {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  userId    String?
  name      String
  user      User?    @relation(fields: [userId], references: [id])
  songs     Song[]
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`can be disabled: Rentals Platform 1`] = `
Object {
  "content": "// Schema for a fictional hotel
// This hotel tracks rooms, user reservations & reviews

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
  provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model User {
  id           String        @id @default(uuid())
  createdAt    DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  email        String        @unique
  name         String
  reservations Reservation[]
  reviews      Review[]
}

model Reservation {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  userId    String?
  startDate DateTime
  endDate   DateTime
  price     Int
  total     Int
  roomId    String?
  room      Room?    @relation(fields: [roomId], references: [id])
  user      User?    @relation(fields: [userId], references: [id])
}

model Room {
  id                 String        @id @default(uuid())
  createdAt          DateTime      @default(now())
  updatedAt          DateTime      @updatedAt
  totalOccupancy     Int           @default(5)
  totalBedrooms      Int           @default(5)
  totalBathrooms     Int           @default(5)
  summary            String?
  address            String
  isSmoking          Boolean       @default(false)
  hasTV              Boolean       @default(true)
  hasAirConditioning Boolean       @default(true)
  hasHeating         Boolean       @default(true)
  hasWIFI            Boolean       @default(true)
  price              Int
  media              Media[]
  reservations       Reservation[]
}

model Review {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  rating    Int
  comment   String
  userId    String
  user      User     @relation(fields: [userId], references: [id])
}

model Media {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  roomId    String?
  fileName  String
  room      Room?    @relation(fields: [roomId], references: [id])
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`can be disabled: SaaS 1`] = `
Object {
  "content": "// Schema for an identity service for a SaaS
// This service tracks existing users & their accounts, as well as invites for users who haven't joined yet

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
  provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model User {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String?
  email     String
  accountId String?
  account   Account? @relation(fields: [accountId], references: [id])
}

model Account {
  id                   String   @id @default(uuid())
  createdAt            DateTime @default(now())
  updatedAt            DateTime @updatedAt
  stripeCustomerId     String
  stripeSubscriptionId String
  referrer             String?
  isActive             Boolean
  invites              Invite[]
  users                User[]
}

model Invite {
  id        String   @id @default(uuid())
  dateSent  DateTime @default(now())
  email     String
  accountId String?
  isValid   Boolean  @default(true)
  account   Account? @relation(fields: [accountId], references: [id])
}",
  "path": "prisma/schema.prisma",
}
`;

exports[`can be disabled: URL Shortener 1`] = `
Object {
  "content": "// Schema for a simple URL shortener
// This schema tracks short links by associating them to a user

generator client {
  provider = \\"prisma-client-js\\"
}

datasource db {
  provider = \\"mysql\\"
  url      = env(\\"DATABASE_URL\\")
}

model Link {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  url       String
  shortUrl  String
  userId    String?
  User      User?    @relation(fields: [userId], references: [id])
}

model User {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  name      String?
  email     String
  links     Link[]
}",
  "path": "prisma/schema.prisma",
}
`;
